server/src/main/java/org/candlepin/controller/ActualRefresher.java

```diff
@@ -0,0 +1,1297 @@
+            }
+        }
+
+
+
+
+        protected Map<EntityNode.Type, Map<String, EntityNode>> typeMap;
+
+
+        public NodeMapper() {
+            this.typeMap = new HashMap<>();
+        }
+
+        public boolean hasNode(EntityNode.Type type, String id) {
+            Map<String, EntityNode> nodeMap = this.typeMap.get(type);
+            return nodeMap != null && nodeMap.containsKey(id);
+        }
+
+        public boolean hasNodesOfType(EntityNode.Type type) {
+            Map<String, EntityNode> nodeMap = this.typeMap.get(type);
+            return nodeMap != null;
+        }
+
+        public EntityNode getNode(EntityNode.Type type, String id) {
+            Map<String, EntityNode> nodeMap = this.typeMap.get(type);
+            return nodeMap != null ? nodeMap.get(id) : null;
+        }
+
+        public Set<Map.Entry<String, EntityNode>> getNodesByType(EntityNode.Type type) {
+            Map<String, EntityNode> nodeMap = this.typeMap.get(type);
+            return nodeMap != null ? nodeMap.entrySet() : null;
+        }
+
+        public Stream<EntityNode> streamNodes() {
+            // This is pretty bad, come up with something nicer before final implementation
+            Stream<EntityNode> stream = Stream.empty();
+
+            for (Map<String, EntityNode> idMap : this.typeMap.values()) {
+                stream = Stream.concat(stream, idMap.values().stream());
+            }
+
+            return stream;
+        }
+
+        public EntityNode putNode(EntityNode.Type type, String id, EntityNode node) {
+            if (type == null) {
+                throw new IllegalArgumentException("type is null");
+            }
+
+            if (id == null || id.isEmpty()) {
+                throw new IllegalArgumentException("node id is null or empty");
+            }
+
+            if (node == null) {
+                throw new IllegalArgumentException("node is null");
+            }
+
+            Map<String, EntityNode> nodeMap = this.typeMap.get(type);
+
+            if (nodeMap == null) {
+                nodeMap = new HashMap<>();
+                this.typeMap.put(type, nodeMap);
+            }
+
+            return nodeMap.put(id, node);
+        }
+
+        public Iterator<EntityNode> getRootIterator() {
+            return new RootIterator(this.typeMap.values().iterator());
+        }
+    }
+
+
+    private static class EntityNode<E extends AbstractHibernateObject, I extends ServiceModelInfo> {
+
+        public static enum Type {
+            PRODUCT,
+            CONTENT
+        };
+
+        private final Type type;
+
+        private E existingEntity;
+        private I importedEntity;
+        private E updatedEntity;
+
+        private boolean changed; // Eventually change this to a full set of changes detected
+        private boolean visited;
+
+        private Set<EntityNode> parents;
+        private Set<EntityNode> children;
```

Comment:
Yes and no. The way they're used in this implementation, a node should never be processed or initiated more than once, so the children and parents shouldn't interfere with each other.

However, for the sake of defensive programming and not causing problems in later maintenance, we could add something. It might be a bit difficult without being able to pull information about the specific entity, however. We'd need to give it more information to go on, or change the type of the parameters.